src/realm.js

```diff
@@ -1324,6 +1372,19 @@ export class Realm {
   // there is already an entry for binding.
   recordModifiedProperty(binding: void | PropertyBinding): void {
     if (binding === undefined) return;
+    if (
+      this.isInPureScope() &&
+      this.reportSideEffectCallback !== undefined &&
+      this.createdObjectsTrackedForLeaks !== undefined &&
+      binding.object instanceof ObjectValue &&
```

Comment:
Can this ever be AbstractObjectValue here? If not, this should be an invariant. If we can, then we need to handle it because it can cover up mutations if it is wrapped by an abstract value.

Reply:
Will add those checks